onnxruntime/test/providers/kernel_compute_test_utils.cc

```diff
@@ -55,12 +55,20 @@ void KernelComputeTester::Run(std::unordered_set<int> strided_outputs) {
     }
 #if defined(USE_CUDA) || defined(USE_ROCM)
     if ((provider_ == kCudaExecutionProvider || provider_ == kRocmExecutionProvider) && !data.is_cpu_data_) {
-      OrtValue gpu_value;
       const Tensor& tensor = data.value_.Get<Tensor>();
-      Tensor::InitOrtValue(tensor.DataType(), tensor.Shape(),
-                           execution_providers.Get(ep_type)->CreatePreferredAllocators()[0], gpu_value,
-                           tensor.Strides());
-      ASSERT_STATUS_OK(dtm.CopyTensor(tensor, *gpu_value.GetMutable<Tensor>()));
+
+      Tensor gpu_tensor(tensor.DataType(), tensor.Shape(),
+                        execution_providers.Get(ep_type)->CreatePreferredAllocators()[0]);
+
+      if (const auto strides = tensor.Strides(); !strides.empty()) {
+        gpu_tensor.SetShapeAndStrides(tensor.Shape(), strides);
+      }
```

Comment:
@er3x3 any concerns about this change to not support passing `strides` to the Tensor ctor that allocates a buffer? It's only test code that uses this facility so doesn't make sense to me to pollute the Tensor ctor when SetShapeAndStrides can be used (and SetShapeAndStrides is used by production code).

Reply:
It depends on how far we want to support the strided tensor. I actually had some features using this but not pushed yet. For example, ConstantOfShape, currently we will allocate full size of memory, but we can use this strided parameter to control and allocate only one-element memory if following nodes can support strided tensor. I think previously the eager mode also wanted this? Or maybe they were switching to another way.

I think it's OK to remove the strides for now if no real code uses it and currently ORT just doesn't support strided tensor well. But if in the future we want to explore on this way, I am guessing we will add them back.

Reply:
I'm not removing the strides overall - just the additional argument to the Tensor ctor where the Tensor is going to do the allocation as it's not clear when we would need that in production code. The typical (AFAIK) usage of creating a Tensor with strides that applies to an existing buffer (i.e. the Tensor is not going to do the allocation and p_data is provided with the existing buffer) is still supported. 

e.g. for ConstantOfShape wouldn't the constant value come from an existing buffer for the tensor attribute of node (ConstantOfShapeBase::GetValuePtr), and in ConstantOfShape::Compute we'd create a Tensor, provide that existing buffer, and create the correct `strides` value for the ConstantOfShape input argument?

Reply:
I get your point, you are right that when providing existing buffer, the strides can be set by SetShapeAndStrides, then the user needs to do this explicitly in two steps.

For ConstantOfShape, the constant value is in attribute, and it's optional. At least sharing buffer will not work on CUDA, the kernel actually needs to allocate new buffer and fill the data.

Reply:
Got it. If you think the ConstantOfShape optimization will be added in the near future I can leave the `strides` parameter there.

BTW is there documentation somewhere on how the strided tensor works? I don't quite understand the size calculation using the actual dim size - 1. 

```c++
int64_t GetSizeFromStrides(const TensorShape& shape, gsl::span<const int64_t> strides) {
  SafeInt<int64_t> size = 1;
  for (size_t dim = 0; dim < shape.NumDimensions(); ++dim) {
    if (shape[dim] == 0) {
      size = 0;
      break;
    }
    size += strides[dim] * (shape[dim] - 1);
  }
  return size;
}
```

Reply:
Since there is no model that can benefit a lot from the ConstantOfShape optimization, I will not push this in near future, so I guess removing it for now is OK. I can add it back if necessary.

There is design doc in the team's OneDrive under ONNX Training folder for strided tensor.

The algorithm of size compute using strides is a common one, you can find similar code in PyTorch repo. The idea is to compute the offset of last element in the buffer, say its x, then the total size is x + 1 (you can see that the size is initialized with 1). Take a normal tensor with shape [2,3,3] as example. the strides is [9,3,1], then the offset of the last element of the buffer should be 1*9+2*3+2*1=17, then the buffer size is 17+1=18=2*3*3.